fix(k8s): avoid issues with NFS provisioner on node/pod eviction

edvald opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the https://github.com/garden-io/garden/blob/master/CONTRIBUTING.md file.
2. Please label this pull request according to what type of issue you are addressing (see "What type of PR is this?" below)
3. Ensure you have added or ran the appropriate tests for your PR.
4. If the PR is unfinished, add `WIP:` at the beginning of the title or use the Github Draft PR feature.
5. Please add at least two reviewers to the PR. Currently active maintainers are: @edvald, @thsig, @eysi09,  @ellenkorbes, @10ko.
-->

**What this PR does / why we need it**:

Enabling persistence hould avoid problems when pods
are evicted, e.g. due to resource constraints or node upgrades.

We're making the transition relatively seamless by creating a new
storage class, and switching from StatefulSet to Deployment (which
makes the underlying volume easier to manage and manipulate).

This needed a fork of the Helm chart, so we now need to track that
manually.

**Special notes for reviewer**:

I suggest trying this out on both a fresh cluster, and one that already had a previous version installed. Run `garden plugins kubernetes cluster-init` and afterwards make sure building works nicely, and the `garden-system` namespace looks as expected.

stevenpall commented:
@edvald I was testing out this branch, and one issue I noticed is that if you specify an existing storageClass (e.g. `efs`) under `providers[].storage.sync.StorageClass`, the nfs-provisioner seems to attempt to _create_ said storageClass. This operation fails due to the naming conflict, and the provisioner then proceeds to _remove_ the storageClass as a cleanup operation. This is obviously problematic.

If a different storageClass name is specified (e.g. `efs-test`), the nfs-provisioner successfully creates a storageClass with said name and of type `cluster.local/garden-nfs-provisioner-v2`. That said, this storageClass obviously does not utilize the remote storage (EFS in this case) and instead connects to the persistent volume-backed NFS endpoint created by the nfs-provisioner. In this case I would expect `cluster-init` to throw an error instead since the specified storageClass does not exist.

stevenpall commented:
Also, it seems that when using the default `garden-system-nfs-v2` storageClass, the NFS export does not recover from underlying node failure properly.

The `pvc-` file that should exist under `/exports`in the `nfs-provisioner` pod seemingly does not, and thus the `build-sync` and `docker-daemon` pods cannot mount this and start up.

eysi09 commented:
> On a related note, when uninstall-garden-services is run, the garden-build-sync and garden-docker-daemon pods seem to get stuck in a Terminating state and must be force-removed.

Sounds like this might be related to #1232

eysi09 commented:
@edvald Does this address #1184?

edvald commented:
No, that is a separate issue.

edvald commented:
@stevenpall I believe I've addressed the issues we found. We now persist the data as well, not just the config, since I found no immediately workable way to fix the issue in the provisioner. Note that since you tried this branch previously, you may need to remove the resources created, or simply delete the `garden-system` namespace in the cluster you used and run `cluster-init` again.

The issues with `uninstall-garden-services` relate to the NFS provisioner and are a bit hard for us to mitigate, but it appears they are less likely to occur with the config and data being persisted.

edvald commented on garden-service/src/plugins/kubernetes/commands/cluster-init.ts:
```diff
@@ -34,6 +36,19 @@ export const clusterInit: PluginCommand = {
+        namespace: "garden-system",
+        args: ["delete", "--purge", "garden-nfs-provisioner"],
+      })
+    } catch (_) { }
```
This is expected to fail in most cases, since the transition will happen only once. Wouldn't hurt to add a debug log though.

edvald commented on .circleci/config.yml:
```diff
@@ -380,7 +382,7 @@ jobs:
       - run:
           name: Cleanup
-          command: (kubectl delete --wait=false $(kubectl get ns -o name | grep $env:CIRCLE_BUILD_NUM)) -or $true
+          command: (kubectl delete --wait=false demo-project-testing-$CIRCLE_BUILD_NUM -or $true)
```
Ah yes, good catch. Since you've looked at it, could you paste a correctly formulated line for me? 😬

eysi09 commented on .circleci/config.yml:
```diff
@@ -312,10 +313,11 @@ jobs:
       - run:
           name: Cleanup
-          command: kubectl delete --wait=false $(kubectl get ns -o name | grep $CIRCLE_BUILD_NUM) || true
+          command: kubectl delete --wait=false demo-project-testing-$CIRCLE_BUILD_NUM-dist || true
```
`ns` or `namespace` is missing from the command.

eysi09 replied:
This also won't delete the metadata namespace.

edvald commented:
@eysi09 this should be good to go now
